File: ios-native/Sprout/Design/Feel.swift
import CoreHaptics
import UIKit

/// Отклик в руке и на слух. Весь отклик — рисунки `CoreHaptics`, даже щелчок
/// выбора: системные `UIFeedbackGenerator` не слушают ползунок силы. Рисунки
/// — в модели, см. `Pulse`, `Rain` и `Knock`; звуки — см. `Chime`. Выбор и
/// всходы без звука: выбор щёлкает сам, а всходы идут и при запуске.
@MainActor
enum Feel {
    /// Капли каждый раз новые — зерно случайное.
    static func water() {
        Engine.shared.play(Rain.drops(over: Motion.cheerSeconds,
                                      seed: .random(in: .min ... .max)),
                           long: true)
        Chime.pour.play()
    }

    static func planted() {
        Engine.shared.play(.bloom, seconds: 0.55)
        Chime.plant.play()
    }

    static func sprout() {
        Engine.shared.play(.sprout, seconds: Motion.bloomSeconds)
    }

    static func frenzy() {
        Engine.shared.play(.frenzy, seconds: Frolic.seconds)
        Chime.frolic.play()
    }

    /// Удалили — растение или запись из истории.
    static func toss() {
        Engine.shared.play(Knock.toss)
        Chime.toss.play()
    }

    /// Отменили удаление или полив.
    static func back() {
        Engine.shared.play(Knock.back)
        Chime.undo.play()
    }

    /// Выбрали из нескольких — щелчок барабана. Переключателям и системным
    /// меню он не нужен: они отзываются сами.
    static func pick() {
        Engine.shared.play(Knock.pick)
    }

    static func done() {
        Engine.shared.play(Knock.done)
        Chime.save.play()
    }

    static func wrong() {
        Engine.shared.play(Knock.wrong)
        Chime.wrong.play()
    }

    /// Проба силы с ползунка — без звука: пробуют руку.
    static func sample() {
        Engine.shared.play(Knock.done)
    }

    /// Комната сменилась — мягкий глубокий толчок, как у барабана с тяжёлой
    /// осью: тупой, без щелчка. `weight` — сила.
    static func turn(_ weight: Double = 0.7) {
        Engine.shared.thud(weight)
    }

    /// Тянут к «Новой комнате» — гул, который растёт вместе с оттяжкой;
    /// ноль — тишина.
    static func pull(_ share: Double) {
        Engine.shared.pull(share)
    }
}

/// Taptic Engine — один на приложение: запуск движка на каждый полив стоил бы
/// дорого. Засыпает он сам, а `start` на ходу ничего не стоит.
@MainActor
private final class Engine {
    static let shared = Engine()

    /// На iPad и в симуляторе его нет — отклик просто не играет.
    private let supported = CHHapticEngine.capabilitiesForHardware()
        .supportsHaptics

    private var engine: CHHapticEngine?

    /// Новый долгий отклик обрывает прежний: две волны капель читались бы
    /// кашей. Как и новая волна — см. `Cheer.now`. Короткие поверх долгих не
    /// мешают и его не обрывают.
    private var playing: CHHapticPatternPlayer?

    /// Гул оттяжки — один, сила меняется на ходу.
    private var pulling: CHHapticAdvancedPatternPlayer?

    private init() {}

    private var strength: Double { Settings.shared.hapticStrength }

    /// Толчок: удар почти без остроты и короткий низкий гул следом — рука
    /// читает его глубиной, а не щелчком.
    func thud(_ weight: Double) {
        guard supported, strength > 0 else { return }
        let deep = Float(Pulse.scaled(weight, by: strength))
        do {
            let pattern = try CHHapticPattern(events: [
                CHHapticEvent(eventType: .hapticTransient, parameters: [
                    CHHapticEventParameter(parameterID: .hapticIntensity,
                                           value: deep),
                    CHHapticEventParameter(parameterID: .hapticSharpness,
                                           value: 0.1),
                ], relativeTime: 0),
                CHHapticEvent(eventType: .hapticContinuous, parameters: [
                    CHHapticEventParameter(parameterID: .hapticIntensity,
                                           value: deep * 0.55),
                    CHHapticEventParameter(parameterID: .hapticSharpness,
                                           value: 0.02),
                ], relativeTime: 0, duration: 0.09),
            ], parameters: [])
            try alive().makePlayer(with: pattern)
                .start(atTime: CHHapticTimeImmediate)
        } catch {
            engine = nil
        }
    }

    /// Гул оттяжки: заводится с первой долей, дальше только меняет силу.
    /// Кривая круче к концу — последние сантиметры чувствуются сильнее.
    func pull(_ share: Double) {
        guard supported else { return }
        let level = Float(Pulse.scaled(pow(min(max(share, 0), 1), 1.4),
                                       by: strength))
        guard level > 0.02 else {
            try? pulling?.stop(atTime: CHHapticTimeImmediate)
            pulling = nil
            return
        }
        do {
            if pulling == nil {
                let pattern = try CHHapticPattern(events: [
                    CHHapticEvent(eventType: .hapticContinuous, parameters: [
                        CHHapticEventParameter(parameterID: .hapticIntensity,
                                               value: 1),
                        CHHapticEventParameter(parameterID: .hapticSharpness,
                                               value: 0.04),
                    ], relativeTime: 0, duration: 30),
                ], parameters: [])
                let player = try alive().makeAdvancedPlayer(with: pattern)
                try player.start(atTime: CHHapticTimeImmediate)
                pulling = player
            }
            try pulling?.sendParameters([
                CHHapticDynamicParameter(parameterID: .hapticIntensityControl,
                                         value: level, relativeTime: 0),
            ], atTime: CHHapticTimeImmediate)
        } catch {
            pulling = nil
        }
    }

    func play(_ pulse: Pulse, seconds: Double) {
        var taps = pulse.taps(over: seconds)
        if pulse.strike > 0 {
            taps.insert(Tap(at: 0, strength: pulse.strike,
                            edge: pulse.strikeEdge), at: 0)
        }
        if pulse.finish > 0 {
            taps.append(Tap(at: seconds, strength: pulse.finish,
                            edge: pulse.finishEdge))
        }
        play(taps, long: true,
             hum: pulse.hum > 0 ? (pulse: pulse, seconds: seconds) : nil)
    }

    func play(_ taps: [Tap], long: Bool = false,
              hum: (pulse: Pulse, seconds: Double)? = nil) {
        guard supported, strength > 0, !taps.isEmpty else { return }
        do {
            let engine = try alive()
            let player = try engine.makePlayer(with: try pattern(taps, hum: hum))
            if long { try? playing?.stop(atTime: CHHapticTimeImmediate) }
            try player.start(atTime: CHHapticTimeImmediate)
            if long { playing = player }
        } catch {
            // Отклик — украшение, падать из-за него нечего: движок соберётся
            // заново.
            engine = nil
            playing = nil
        }
    }

    private func alive() throws -> CHHapticEngine {
        if let engine {
            // Мог уснуть сам, а на спящем проигрывание не начнётся.
            try engine.start()
            return engine
        }
        let made = try CHHapticEngine()
        // Звука нет — так движок не трогает звуковую сессию приложения.
        made.playsHapticsOnly = true
        made.isAutoShutdownEnabled = true
        // Сброшенный системой (например, после звонка) движок надо поднять
        // заново, иначе он замолчит навсегда.
        made.resetHandler = { [weak made] in try? made?.start() }
        try made.start()
        engine = made
        return made
    }

    /// Тычки — мгновенные события. Гул — одно долгое событие в полную силу,
    /// форму которому задаёт кривая `hapticIntensityControl`: событие с
    /// заданной силой угасать не умеет.
    private func pattern(_ taps: [Tap],
                         hum: (pulse: Pulse, seconds: Double)?) throws
        -> CHHapticPattern {
        var events: [CHHapticEvent] = taps.map { tap in
            CHHapticEvent(
                eventType: .hapticTransient,
                parameters: [
                    CHHapticEventParameter(
                        parameterID: .hapticIntensity,
                        value: Float(Pulse.scaled(tap.strength,
                                                  by: strength))),
                    CHHapticEventParameter(parameterID: .hapticSharpness,
                                           value: Float(tap.edge)),
                ],
                relativeTime: tap.at)
        }
        var curves: [CHHapticParameterCurve] = []
        if let hum {
            events.append(CHHapticEvent(
                eventType: .hapticContinuous,
                parameters: [
                    CHHapticEventParameter(parameterID: .hapticIntensity,
                                           value: 1),
                    CHHapticEventParameter(parameterID: .hapticSharpness,
                                           value: Float(hum.pulse.edge)),
                ],
                relativeTime: 0,
                duration: hum.seconds))
            curves.append(CHHapticParameterCurve(
                parameterID: .hapticIntensityControl,
                controlPoints: hum.pulse.envelope.map {
                    CHHapticParameterCurve.ControlPoint(
                        relativeTime: $0.at * hum.seconds,
                        value: Float(Pulse.scaled(
                            $0.strength * hum.pulse.hum, by: strength)))
                },
                relativeTime: 0))
        }
        return try CHHapticPattern(events: events, parameterCurves: curves)
    }
}

File: android/app/src/main/kotlin/com/ledro6/sprout/platform/Feel.kt
package com.ledro6.sprout.platform

import android.annotation.SuppressLint
import android.content.Context
import android.media.AudioAttributes
import android.media.AudioManager
import android.media.SoundPool
import android.os.Build
import android.os.VibrationAttributes
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import com.ledro6.sprout.R
import com.ledro6.sprout.design.Motion
import com.ledro6.sprout.model.Frolic
import com.ledro6.sprout.model.Knock
import com.ledro6.sprout.model.Pulse
import com.ledro6.sprout.model.Rain
import com.ledro6.sprout.model.Settings
import com.ledro6.sprout.model.Tap
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.roundToInt
import kotlin.random.Random

/**
 * Отклик в руке и звук — вместе, но со своими настройками. Рисунки те же,
 * что на iPhone (`Pulse`, `Rain`, `Knock`); играет их вибромотор телефона:
 * на Android 11+ — примитивами «щелчок» и «тик», которые сами подстраиваются
 * под мотор, на старых — ступенчатой волной силы. Отклик — украшение: любой
 * сбой мотора молча пропускается.
 */
object Feel {
    private var vibrator: Vibrator? = null
    private var settings: Settings? = null

    fun open(context: Context, settings: Settings) {
        this.settings = settings
        vibrator = runCatching {
            if (Build.VERSION.SDK_INT >= 31) {
                (context.getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager).defaultVibrator
            } else {
                @Suppress("DEPRECATION")
                context.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
            }
        }.getOrNull()?.takeIf { it.hasVibrator() }
        Chimes.open(context, settings)
    }

    private val strength: Double get() = settings?.hapticStrength ?: 0.0

    fun water() {
        play(Rain.drops(Motion.CHEER_SECONDS, Random.nextLong()))
        Chimes.play(Chime.POUR)
    }

    fun planted() {
        play(Pulse.bloom, 0.55)
        Chimes.play(Chime.PLANT)
    }

    fun sprout() = play(Pulse.sprout, Motion.BLOOM_SECONDS)

    fun frenzy() {
        play(Pulse.frenzy, Frolic.SECONDS)
        Chimes.play(Chime.FROLIC)
    }

    fun toss() {
        play(Knock.toss)
        Chimes.play(Chime.TOSS)
    }

    fun back() {
        play(Knock.back)
        Chimes.play(Chime.UNDO)
    }

    fun pick() = play(Knock.pick)

    fun done() {
        play(Knock.done)
        Chimes.play(Chime.SAVE)
    }

    fun wrong() {
        play(Knock.wrong)
        Chimes.play(Chime.WRONG)
    }

    /** Показать силу с ползунка — без звука. */
    fun sample() = play(Knock.done)

    /** Глухой толчок листания — как барабан. */
    fun turn(weight: Double = 0.7) = play(listOf(Tap(0.0, weight, 0.1)))

    private var pulled = 0.0

    /** Гул оттяжки к «Новой комнате»: растёт вместе с ней. */
    @SuppressLint("MissingPermission")
    fun pull(share: Double) {
        val motor = vibrator ?: return
        val level = Pulse.scaled(min(max(share, 0.0), 1.0).pow(1.4), strength)
        if (level <= 0.02) {
            if (pulled > 0) runCatching { motor.cancel() }
            pulled = 0.0
            return
        }
        if (kotlin.math.abs(level - pulled) < 0.08) return
        pulled = level
        if (!motor.hasAmplitudeControl()) return
        runCatching {
            motor.vibrate(VibrationEffect.createOneShot(400, (level * 120).roundToInt().coerceIn(1, 255)))
        }
    }

    fun play(pulse: Pulse, seconds: Double) {
        val taps = pulse.taps(seconds).toMutableList()
        if (pulse.strike > 0) taps.add(0, Tap(0.0, pulse.strike, pulse.strikeEdge))
        if (pulse.finish > 0) taps.add(Tap(seconds, pulse.finish, pulse.finishEdge))
        play(taps)
    }

    @SuppressLint("MissingPermission")
    fun play(taps: List<Tap>) {
        val motor = vibrator ?: return
        if (strength <= 0 || taps.isEmpty()) return
        val effect = runCatching { effect(motor, taps) }.getOrNull() ?: return
        runCatching {
            if (Build.VERSION.SDK_INT >= 33) {
                motor.vibrate(effect, VibrationAttributes.createForUsage(VibrationAttributes.USAGE_TOUCH))
            } else {
                @Suppress("DEPRECATION")
                motor.vibrate(effect, AudioAttributes.Builder().setUsage(AudioAttributes.USAGE_ASSISTANCE_SONIFICATION).build())
            }
        }
    }

    /** Рисунок в эффект мотора — самым богатым способом, который мотор умеет. */
    fun effect(motor: Vibrator, taps: List<Tap>): VibrationEffect? {
        val sorted = taps.sortedBy { it.at }
        if (Build.VERSION.SDK_INT >= 30 &&
            motor.areAllPrimitivesSupported(VibrationEffect.Composition.PRIMITIVE_CLICK, VibrationEffect.Composition.PRIMITIVE_TICK)
        ) {
            val low = Build.VERSION.SDK_INT >= 31 &&
                motor.areAllPrimitivesSupported(VibrationEffect.Composition.PRIMITIVE_LOW_TICK)
            val composition = VibrationEffect.startComposition()
            var last = 0.0
            var added = 0
            for (tap in sorted) {
                val scale = Pulse.scaled(tap.strength, strength).toFloat()
                if (scale <= 0f) continue
                val primitive = when {
                    tap.edge >= 0.55 -> VibrationEffect.Composition.PRIMITIVE_CLICK
                    tap.edge < 0.25 && low -> VibrationEffect.Composition.PRIMITIVE_LOW_TICK
                    else -> VibrationEffect.Composition.PRIMITIVE_TICK
                }
                val delay = ((tap.at - last) * 1000).roundToInt().coerceAtLeast(0)
                composition.addPrimitive(primitive, scale.coerceIn(0f, 1f), delay)
                last = tap.at
                added += 1
            }
            return if (added == 0) null else composition.compose()
        }
        return waveform(sorted, motor.hasAmplitudeControl())
    }

    /** Ступенчатая волна: тычок — короткий всплеск силы, между ними тишина. */
    fun waveform(taps: List<Tap>, amplitudes: Boolean): VibrationEffect? {
        val timings = mutableListOf<Long>()
        val levels = mutableListOf<Int>()
        var clock = 0L
        for (tap in taps) {
            val force = Pulse.scaled(tap.strength, strength)
            if (force <= 0) continue
            val start = (tap.at * 1000).toLong()
            // Резкий — короче: так он читается щелчком, а не гулом.
            val length = (26 - 14 * tap.edge).roundToInt().toLong()
            if (start > clock) {
                timings.add(start - clock)
                levels.add(0)
                clock = start
            }
            timings.add(length)
            levels.add((force * 255).roundToInt().coerceIn(1, 255))
            clock += length
        }
        if (timings.isEmpty()) return null
        return if (amplitudes) {
            VibrationEffect.createWaveform(timings.toLongArray(), levels.toIntArray(), -1)
        } else {
            // Мотор без силы: только «вкл/выкл», слабые тычки пропускаем.
            val on = timings.indices.map { if (levels[it] > 60) timings[it] else 0L }
            val pattern = mutableListOf(0L)
            for ((i, time) in timings.withIndex()) {
                if (levels[i] == 0 || on[i] == 0L) pattern[pattern.lastIndex] += time
                else {
                    pattern.add(time)
                    pattern.add(0L)
                }
            }
            if (pattern.size < 2) null else VibrationEffect.createWaveform(pattern.toLongArray(), -1)
        }
    }
}

/** Звуки приложения — те же файлы, что на iPhone. */
enum class Chime(val res: Int) {
    POUR(R.raw.chime_pour), PLANT(R.raw.chime_plant), TOSS(R.raw.chime_toss), UNDO(R.raw.chime_undo),
    SAVE(R.raw.chime_save), WRONG(R.raw.chime_wrong), FROLIC(R.raw.chime_frolic), STREAM(R.raw.chime_stream),
}

/**
 * Звуки — через `SoundPool`: короткие, без задержки на разбор файла. Как на
 * iPhone, беззвучный режим телефона их глушит: в режиме «без звука» и
 * «вибрация» приложение молчит.
 */
object Chimes {
    private var pool: SoundPool? = null
    private val ids = HashMap<Chime, Int>()
    private val ready = HashSet<Int>()
    private var audio: AudioManager? = null
    private var settings: Settings? = null

    fun open(context: Context, settings: Settings) {
        this.settings = settings
        if (pool != null) return
        audio = context.getSystemService(Context.AUDIO_SERVICE) as? AudioManager
        val made = runCatching {
            SoundPool.Builder()
                .setMaxStreams(4)
                .setAudioAttributes(
                    AudioAttributes.Builder()
                        .setUsage(AudioAttributes.USAGE_ASSISTANCE_SONIFICATION)
                        .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                        .build(),
                )
                .build()
        }.getOrNull() ?: return
        made.setOnLoadCompleteListener { _, id, status -> if (status == 0) ready.add(id) }
        pool = made
        for (chime in Chime.entries) {
            runCatching { made.load(context, chime.res, 1) }.getOrNull()?.let { ids[chime] = it }
        }
    }

    fun play(chime: Chime) {
        if (settings?.sounds != true) return
        if (audio?.ringerMode != AudioManager.RINGER_MODE_NORMAL) return
        val id = ids[chime] ?: return
        if (id !in ready) return
        runCatching { pool?.play(id, 0.8f, 0.8f, 1, 0, 1f) }
    }
}
